src/main/java/org/wso2/sample/identity/oauth2/grant/ed25519/X509GrantHandler.java

```diff
@@ -40,83 +42,94 @@
-        if(subjectDN != null) {
-            username = subjectDN;
-            log.debug("Username is retrieved from Subject DN : " + username);
-        } else {
-            if (certParam != null) {
-                username = getUserFromSSHCert(certParam);
-            }
+        // Extract 'Principal' from the presented certificate.
+        if (certificateParameterValue != null) {
+            username = ExtractUserFromSSHCertificate(certificateParameterValue);
         }
 
-        UserStoreManager userStoreManager;
-        int tenantId = IdentityTenantUtil.getTenantId(oAuthTokenReqMessageContext.getOauth2AccessTokenReqDTO().getTenantDomain());
+        if (username == null) {
+            throw new IdentityOAuth2Exception("Principal was not found in the provided certificate.");
+        }
+
+        UserStoreManager userStoreManager = getUserStoreManager(oAuthTokenReqMessageContext.getOauth2AccessTokenReqDTO()
+                .getTenantDomain());
+
         try {
-            userStoreManager = GrantHandlerServiceComponent.getRealmService().
-                    getTenantUserRealm(tenantId).getUserStoreManager();
-            if (username != null && userStoreManager != null && !userStoreManager.isExistingUser(username)) {
+            // Check whether the 'Principal' is an existing user.
+            if (userStoreManager != null && !userStoreManager.isExistingUser(username)) {
                 throw new IdentityOAuth2Exception("User: " + username + " does not exist.");
             }
         } catch (UserStoreException e) {
-            throw new IdentityOAuth2Exception("Error occurred while getting user store manager", e);
+            throw new IdentityOAuth2Exception("Error occurred while performing 'isExistingUser' operation.", e);
         }
 
-        // if valid set authorized mobile number as grant user
         AuthenticatedUser user = OAuth2Util.getUserFromUserName(username);
         user.setAuthenticatedSubjectIdentifier(user.toString());
         oAuthTokenReqMessageContext.setAuthorizedUser(user);
         oAuthTokenReqMessageContext.setScope(oAuthTokenReqMessageContext.getOauth2AccessTokenReqDTO().getScope());
+
         return true;
     }
 
-    private String getUserFromSSHCert(String sshCertificate) {
+    private UserStoreManager getUserStoreManager(String tenantDomain)
+            throws IdentityOAuth2Exception {
+
+        int tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
+
+        try {
+            return GrantHandlerServiceComponent.getRealmService().getTenantUserRealm(tenantId).getUserStoreManager();
+
+        } catch (UserStoreException e) {
+            throw new IdentityOAuth2Exception("Error occurred while getting user store manager", e);
+        }
+    }
+
+    private String ExtractUserFromSSHCertificate(String sshCertificate) {
 
         String[] split = sshCertificate.trim().split("\\s+");
         byte[] encodedCert = java.util.Base64.getMimeDecoder().decode(split[1].getBytes(UTF_8));
```

Comment:
let's check `split.length == 2`

Reply:
Resolved in https://github.com/gayashanbc/x509-ed25519-grant-handler/pull/3/commits/d01c3a2846a13850ced1a2e93b47bb46687eb8e1